### PR TITLE
Make MoE dispatch/MLP expert-axis batch sharding configurable (fix Mixtral EP throughput)

### PR DESCRIPTION
**Pull request by gulsumgudukbay**
# Description
PR #4007 added `'expert'` to `activation_batch_moe` to fix a DeepSeek MoE throughput regression. That change is applied to the post-dispatch core MoE activations (`dispatch_axis`/`mlp_axis`) as well, where the tensor's expert dim `E` is already
sharded via `activation_exp`. 
For models with few large experts (e.g. Mixtral), mapping the batch dim onto `'expert'` too double-maps two tensor dims onto one mesh axis, so GSPMD abandons the expert-local layout and falls back to FSDP-style AllGather+ReduceScatter instead of expert-parallel AllToAll, creatiing a large throughput regression under single-node expert parallelism (`ici_expert_parallelism=-1`).
This PR adds a config flag `moe_dispatch_no_expert_sharding` (default `false`) that selects, for the training dispatch/MLP batch axis only, a new `activation_batch_no_exp` rule (`['data','fsdp','fsdp_transpose']`, i.e. without `'expert'`). Mixtral-8x7b uses it as **true**.
This is the per-model knob anticipated in the #4007 review discussion (sharding core MoE activations by the expert physical axes rather than the batch dimension), without changing the default for any other model.

## Behavior
- Default `false` is byte-identical to current `main` for every model.
- Only `mixtral-8x7b` opts in; no other config inherits it.
- For `mixtral-8x7b`, the flag only changes sharding when the `expert` mesh axis size > 1  (expert parallelism active). When `expert` is size 1 (TPU/FSDP-primary, or non-expert parallelism GPU) the two axis rules are identical, so those paths are unaffected.

## Scope (why only the training dispatch/MLP axes)
- `mask_axes` is intentionally left on `activation_batch_moe`: switching it too regresses  Mixtral bs=11 (~10,900 -> ~8,400 tok/s/device, train-step all-to-all 5 -> 3).
- The inference-mode `dispatch_axis`/`mlp_axis` are left unchanged: the inference path is already expert-parallel with `activation_batch_moe` (no FSDP fallback) at both prefill (bs=1: all-to-all=290, all-gather=0) and decode (bs=8: all-to-all=193, all-gather=0), so the regression does not occur there. The change is scoped to the training path.

# Tests

Measured on 1x MI355X node (8 GPUs), JAX 0.9.1, `ici_expert_parallelism=-1`, `capacity_factor>0`:
| Model | bs | tok/s/device (before) | tok/s/device (this PR) | train-step all-to-all |
|---|---|---|---|---|
| Mixtral-8x7b | 11 | ~7,400 | ~10,900 | 0 -> 5 |
| Mixtral-8x7b | 6  | ~7,500 | ~11,400 | 0 -> 5 |
| DeepSeek-v2-lite-16b | 8 | ~17,800 | ~17,800 (unchanged) | 0 (unchanged) |

Mixtral recovers ~47% throughput via restored expert-parallel AllToAll; DeepSeek (flag off by default) is unchanged.

Added a unit regression test `test_moe_dispatch_keeps_expert_on_expert_dim` (`tests/unit/moe_test.py`), parametrized over `mixtral-8x7b`, `mixtral-8x22b`, and `deepseek3-671b`: when `moe_dispatch_no_expert_sharding` is set it asserts the expert dim is sharded by the `expert` mesh axis and the batch dim is not (guarding the expert-parallel dispatch/MLP sharding). Passing locally.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/models/mixtral-8x7b.yml [hunk 1]
```diff
@@ -31,3 +31,6 @@ num_experts: 8
 num_experts_per_tok: 2
 rope_max_timescale: 1_000_000
 decoder_block: "mixtral"
+# Few large experts: keep the MoE dispatch/MLP GEMM expert-parallel (AllToAll) rather than
+# letting the batch dim double-map onto 'expert' and fall back to FSDP-style collectives.
+moe_dispatch_no_expert_sharding: true
```
Comment (mfrancepillois) [inline src/maxtext/configs/models/mixtral-8x7b.yml]: This line should probably be added to the `mixtral-8x22b.yml` as well

File: src/maxtext/layers/moe.py [hunk 2]
```diff
@@ -2176,15 +2176,20 @@ def dense_matmul(
...
         mask_axes = ("activation_batch_moe", "activation_norm_length_moe", None, None)
+        # Dispatch/MLP are already expert-sharded via "activation_exp"; with no_expert_sharding the
+        # batch dim drops 'expert' so the GEMM stays expert-parallel (AllToAll) instead of FSDP.
+        moe_dispatch_batch_axis = (
+            "activation_batch_no_exp" if self.config.moe_dispatch_no_expert_sharding else "activation_batch_moe"
+        )
         dispatch_axis = (
             "activation_exp",
-            "activation_batch_moe",
+            moe_dispatch_batch_axis,
```
Comment (mfrancepillois) [inline src/maxtext/layers/moe.py]: Should `activation_batch_moe` also be replaced with `moe_dispatch_batch_axis` in the `dispatch_axis` and the `mlp_axis` below?

**Review by i-chaochen — COMMENTED** [review]
can we add test into [moe_test.py](https://github.com/AI-Hypercomputer/maxtext/blob/main/tests/unit/moe_test.py) for `activation_batch_no_exp` and `moe_dispatch_batch_axis` ? a pure logical check to make sure batch doesn't steal `expert`

**Comment by gulsumgudukbay** [comment]
> can we add test into [moe_test.py](https://github.com/AI-Hypercomputer/maxtext/blob/main/tests/unit/moe_test.py) for `activation_batch_no_exp` and `moe_dispatch_batch_axis` ? a pure logical check to make sure batch doesn't steal `expert`

you beat me! I just pushed it seconds after this comment

File: src/maxtext/configs/models/mixtral-8x7b.yml (hunk 1 above)
Comment (gulsumgudukbay) [reply src/maxtext/configs/models/mixtral-8x7b.yml]: added, thanks for the reminder. 
mixtral-8x22b has the same geometry (8 experts, moe_mlp_dim=16384, decoder_block: mixtral), so it hits the identical batch-dim/expert double-mapping under EP. Added moe_dispatch_no_expert_sharding: true, and it's covered by the new regression test (...[mixtral-8x22b] passes).

File: src/maxtext/layers/moe.py (hunk 2 above)
Comment (gulsumgudukbay) [reply src/maxtext/layers/moe.py]: I benchmarked replacing activation_batch_moe  with moe_dispatch_batch_axis in other sites like dispatch_axis and mlp_axis and it was not beneficial. 

the inference path is already expert-parallel at both prefill bs=1 (jit_prefill_aot: a2a=290, all-gather=0) and decode bs=8 (jit_generate_aot: a2a=193, all-gather=0), no FSDP fallback, unlike unpatched training (a2a=0). So gating the inference dispatch_axis/mlp_axis isn't needed; it only lowers the a2a count without fixing a fallback. I've kept the change scoped to the training path.
I also confirmed earlier that gating training mask_axes regresses throughput ~10.9k to 8.4k, so that's left as-is too.

**Comment by yeandy** [comment]
I tested across many models on MI355 single node. Mixtral regression is resolved without any impact to deepseek-v2-lite. 

Seeing slight regression to llama2-7b, not suspecting that this change is the cause.

File: tests/unit/moe_test.py [hunk 3]
```diff
@@ -1534,5 +1534,38 @@ def test_prefused_vs_sparse_softmax(self):
     self.assertIsNone(bias_updates)
 
 
+@pytest.mark.parametrize("model_name", ["mixtral-8x7b", "mixtral-8x22b", "deepseek3-671b"])
```
Comment (yeandy) [inline tests/unit/moe_test.py]: do we need `deepseek3-671` in the test if `moe_dispatch_no_expert_sharding` is not set to true for this model? we will never hit the `if` conditional, so the test won't fail or anything, but we're also not `assert`ing or testing anything else for models that don't set `moe_dispatch_no_expert_sharding` to true.

Comment (gulsumgudukbay) [reply tests/unit/moe_test.py]: > do we need `deepseek3-671` in the test if `moe_dispatch_no_expert_sharding` is not set to true for this model? we will never hit the `if` conditional, so the test won't fail or anything, but we're also not `assert`ing or testing anything else for models that don't set `moe_dispatch_no_expert_sharding` to true.

you are right deepseek3-671b wasn't asserting anything. Dropped it and now test both states explicitly: flag on (mixtral-8x7b/8x22b) asserts expert is on the expert dim and not the batch dim; a forced flag-off case asserts the default keeps expert on the batch dim. i used Mixtral for the off-case rather than deepseek3-671b since its custom mesh doesn't map activation_exp to expert, so there was nothing stable to assert there.

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/configs/models/mixtral-8x22b.yml [hunk 4]
```diff
@@ -31,3 +31,6 @@ num_experts: 8
 num_experts_per_tok: 2
 rope_max_timescale: 1_000_000
 decoder_block: "mixtral"
+# Few large experts: keep the MoE dispatch/MLP GEMM expert-parallel (AllToAll) rather than
+# letting the batch dim double-map onto 'expert' and fall back to FSDP-style collectives.
+moe_dispatch_no_expert_sharding: true
```
Comment (gobbleturk) [inline src/maxtext/configs/models/mixtral-8x22b.yml]: models.yml files are meant to be core definitions of the model. Implementation details such as sharding are not meant to live in these files. We've made some exceptions for tokenizers since those are a pain to specify and sort of tied to the model

File: src/maxtext/configs/base.yml [hunk 5]
```diff
@@ -218,6 +218,9 @@ load_balance_loss_weight: 0.0 # weight for the load balance loss
 use_random_routing: false # whether to use random routing for debug/test purpose
 use_custom_sort_vjp: true # whether to use a custom VJP sort for efficient backward pass processing in sparse matmul
 use_ring_of_experts: false # whether to use ring of experts for sparse matmul expert parallelism
+# If true, peel the 'expert' mesh axis off the MoE dispatch/MLP batch dim so the expert GEMM
```
Comment (gobbleturk) [inline src/maxtext/configs/base.yml:221]: Can you add this is only relevant for what we call the "dense" case - not our "sparse" shard map case

File: src/maxtext/configs/models/mixtral-8x22b.yml (hunk 4 above)
Comment (gobbleturk) [reply src/maxtext/configs/models/mixtral-8x22b.yml]: Can you instead specify moe_dispatch_no_expert_sharding=True in the commands you care about?

File: src/maxtext/utils/sharding.py [hunk 6]
```diff
@@ -732,6 +732,32 @@ def _remove_fsdp_from_partition_spec(named_sharding):
   return jax.tree.map(_remove_fsdp_from_partition_spec, sharding_tree)
 
 
+def remove_expert_from_partition_spec(pspec, dims_to_peel):
```
Comment (gobbleturk) [inline src/maxtext/utils/sharding.py:735]: this function makes me sad, but I trust the team with help from @NuojCheng can at some point remove the need for these

can keep this function in this PR

In general we should have separate rules for the moe layer vs not the moe layer, which we almost have succesfully refactored into. Any rule named "no_exp" is sad, it should be named logically (as opposed to physically) because it is a logical rule

File: tests/unit/moe_test.py [hunk 7]
```diff
@@ -1534,5 +1535,53 @@ def test_prefused_vs_sparse_softmax(self):
     self.assertIsNone(bias_updates)
 
 
+@pytest.mark.parametrize(
+    "model_name,override_flag",
+    [
+        ("mixtral-8x7b", None),  # opts in via model config (flag on)
```
Comment (gobbleturk) [inline tests/unit/moe_test.py]: can set your new flag in these tests instead of defining them in the model

**Review by Shuwen-Fang — APPROVED** [review]

File: tests/unit/moe_test.py (hunk 7 above)
Comment (gulsumgudukbay) [reply tests/unit/moe_test.py]: addressed in new commit https://github.com/AI-Hypercomputer/maxtext/pull/4179/commits/5709de1951960f63513dde17f21bb10435f80bcb

File: src/maxtext/configs/base.yml (hunk 5 above)
Comment (gulsumgudukbay) [reply src/maxtext/configs/base.yml:221]: Addressed in https://github.com/AI-Hypercomputer/maxtext/pull/4179/commits/5709de1951960f63513dde17f21bb10435f80bcb

File: src/maxtext/configs/models/mixtral-8x22b.yml (hunk 4 above)
Comment (gulsumgudukbay) [reply src/maxtext/configs/models/mixtral-8x22b.yml]: addressed in https://github.com/AI-Hypercomputer/maxtext/pull/4179/commits/5709de1951960f63513dde17f21bb10435f80bcb

**Review by gobbleturk — APPROVED** [review]
